src/App.tsx

```diff
@@ -1,24 +1,239 @@
+    switch (condition) {
+      case ForComletedTodo.Active:
+        return !completed;
+      case ForComletedTodo.Completed:
+        return completed;
+      default:
+        return 1;
+    }
+  }), [condition, todos]);
+
+  const handleDeleteTodo = (todoId: number) => {
+    setInputDisabled(true);
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todoId]);
+
+    return todoService
+      .deleteTodo(todoId)
+      .then((() => {
+        setTodos((prevTodos) => prevTodos.filter(todo => todo.id !== todoId));
+      }))
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+        throw new Error('Unable to delete a todo');
+      })
+      .finally(() => {
+        setInputDisabled(false);
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todoId),
+        );
+      });
+  };
+
+  const handleRenameTodo = async (todo: Todo, newTodoTitle: string) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+    return todoService.updateTodo({
+      id: todo.id,
+      title: newTodoTitle.trim(),
+      userId: todo.userId,
+      completed: todo.completed,
+    })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+        throw new Error('Unable to update a todo');
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+        );
+      });
+  };
+
+  const handleClearCompletedTodos = () => {
+    todos.forEach(todo => {
+      if (todo.completed) {
+        handleDeleteTodo(todo.id);
+      }
+    });
+  };
+
+  const handleToggleTodo = async (todo: Todo) => {
+    try {
+      setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+      const updatedTodo = await todoService.updateTodo({
+        ...todo,
+        completed: !todo.completed,
+      });
+
+      setTodos(prevState => prevState.map(currentTodo => (
+        currentTodo.id !== updatedTodo.id
+          ? currentTodo
+          : updatedTodo
+      )));
+    } catch (error) {
+      setErrorMessage('Unable to update a todo');
+    } finally {
+      setProcessingTodoIds(
+        (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+      );
+    }
+  };
+
+  const handleToggleAllTodos = () => {
+    const activeTodos = todos.filter(todo => !todo.completed);
+
+    if (isAllCompleted) {
+      todos.forEach(handleToggleTodo);
+    } else {
+      activeTodos.forEach(handleToggleTodo);
+    }
+  };
+
+  const handleDelete = async (todoId: number) => {
+    return handleDeleteTodo(todoId);
+  };
+
+  const handleRename = async (todo: Todo, todoTitle: string) => {
+    await handleRenameTodo(todo, todoTitle);
+  };
```

Comment:
Why you need these funcitons? They are just wrappers

Reply:
they are asynchronous and return promises